org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionQuickAssistProcessor.java

```diff
@@ -158,22 +151,23 @@ public ICompletionProposal[] computeQuickAssistProposals(IQuickAssistInvocationC
 		}
 
 		// Get the codeActions
-		CodeActionParams params = prepareCodeActionParams(infos, invocationContext.getOffset(), invocationContext.getLength());
+		CodeActionParams params = prepareCodeActionParams(document, invocationContext.getOffset(), invocationContext.getLength());
 		List<CompletableFuture<Void>> futures = Collections.emptyList();
 		try {
 			// Prevent infinite re-entrance by only computing proposals if there aren't any
 			if (!proposalsRefreshInProgress) {
 				proposals.clear();
-				futures = infos.stream()
-						.map(info -> info.getInitializedLanguageClient()
-								.thenComposeAsync(ls -> ls.getTextDocumentService().codeAction(params)
-										.thenAcceptAsync(actions -> actions.stream().filter(Objects::nonNull)
-												.map(action -> new CodeActionCompletionProposal(action, info))
-												.forEach(p -> processNewProposal(invocationContext, p)))))
-								.toList();
+				// Start all the servers computing actions - each server will append any code actions to the ongoing list of proposals
+				// as a side effect of this request
+				futures = executor.computeAll((w, ls) -> ls.getTextDocumentService()
+						.codeAction(params)
+						.thenAccept(actions -> LanguageServers.streamSafely(actions)
+								.forEach(action -> processNewProposal(invocationContext, new CodeActionCompletionProposal(action, w)))));
 
 				CompletableFuture<?> aggregateFutures = CompletableFuture
 						.allOf(futures.toArray(new CompletableFuture[futures.size()]));
+
+				//
```

Comment:
can we remove this comment?

Reply:
I've enhanced the comment instead to try and clarify better what is going on (it isn't obvious)! See below.